apis/apiextensions/v1/composition_types.go

```diff
@@ -134,6 +134,17 @@ func TypeReferenceTo(gvk schema.GroupVersionKind) TypeReference {
 // ComposedTemplate is used to provide information about how the composed resource
 // should be processed.
 type ComposedTemplate struct {
+	// TODO(negz): TemplateName should be a required field in v2 of this API.
+
+	// A TemplateName uniquely identifies this resource template within its
+	// Composition. Template names are strongly recommended. Either all or no
+	// templates within a Composition must be named. Compositions that use
+	// anonymous templates match templates to composed resources by order, and
+	// must therefore treat the order and length of their spec.resources array
+	// as immutable.
+	// +optional
+	TemplateName *string `json:"templateName,omitempty"`
```

Comment:
Why not use `name` or `id`?

Reply:
You convinced me that `name` was more idiomatic than `id`, but I was worried that `name` alone would easily be mistaken with `metadata.name` in this context so `templateName` seemed like a good way to make it clear that it's the name of the template, not the name of the resulting managed resource.